presto-main/src/main/java/com/facebook/presto/sql/analyzer/MaterializedViewQueryOptimizer.java

```diff
@@ -0,0 +1,321 @@
+    protected Node visitArithmeticBinary(ArithmeticBinaryExpression node, MaterializedViewQueryOptimizerContext context)
+    {
+        return new ArithmeticBinaryExpression(
+                node.getOperator(),
+                (Expression) process(node.getLeft(), context),
+                (Expression) process(node.getRight(), context));
+    }
+
+    @Override
+    protected Node visitIdentifier(Identifier node, MaterializedViewQueryOptimizerContext context)
+    {
+        String baseColumnName = node.getValue();
+        return new Identifier(context.getViewColumnName(baseColumnName), node.isDelimited());
+    }
+
+    @Override
+    protected Node visitFunctionCall(FunctionCall node, MaterializedViewQueryOptimizerContext context)
+    {
+        String functionCall = node.toString();
+        ImmutableList.Builder<Expression> rewrittenArguments = ImmutableList.builder();
+
+        if (context.containsColumnName(functionCall)) {
+            Expression derivedExpression = new Identifier(context.getViewColumnName(functionCall));
+            rewrittenArguments.add(derivedExpression);
+        }
+        else {
+            for (Expression argument : node.getArguments()) {
+                Expression rewrittenArgument = (Expression) process(argument, context);
+                rewrittenArguments.add(rewrittenArgument);
+            }
+        }
+
+        return new FunctionCall(
+                node.getName(),
+                node.getWindow(),
+                node.getFilter(),
+                node.getOrderBy(),
+                node.isDistinct(),
+                node.isIgnoreNulls(),
+                rewrittenArguments.build());
+    }
+
+    // Assuming the current base query applies to this specific materialized view
+    @Override
+    protected Node visitRelation(Relation node, MaterializedViewQueryOptimizerContext context)
+    {
+        Optional<Relation> baseQueryFrom = context.getBaseQueryFrom();
+        if (baseQueryFrom.isPresent() && node.equals(baseQueryFrom.get())) {
+            return context.getMaterializedView();
+        }
+        context.setRewritable(false);
+        return node;
+    }
+
+    @Override
+    protected Node visitLogicalBinaryExpression(LogicalBinaryExpression node, MaterializedViewQueryOptimizerContext context)
+    {
+        return new LogicalBinaryExpression(
+                node.getOperator(),
+                (Expression) process(node.getLeft(), context),
+                (Expression) process(node.getRight(), context));
+    }
+
+    @Override
+    protected Node visitComparisonExpression(ComparisonExpression node, MaterializedViewQueryOptimizerContext context)
+    {
+        return new ComparisonExpression(
+                node.getOperator(),
+                (Expression) process(node.getLeft(), context),
+                (Expression) process(node.getRight(), context));
+    }
+
+    @Override
+    protected Node visitLiteral(Literal node, MaterializedViewQueryOptimizerContext context)
```

Comment:
You perhaps don't need to reimplement visitLiteral(). Check the AstVisitor implementation.